applications/RomApplication/costum_strategies/rom_builder_and_solver.h

```diff
@@ -256,45 +272,66 @@ class ROMBuilderAndSolver : public BuilderAndSolver<TSparseSpace, TDenseSpace, T
         }
     }
 
-    void ProjectToReducedBasis(
-        const TSystemVectorType &rX,
-        ModelPart::NodesContainerType &rNodes,
-        Vector rom_unknowns)
-    {
-        for (const auto &node : rNodes){
-            unsigned int node_aux_id = node.GetValue(AUX_ID);
-            const auto &nodal_rom_basis = node.GetValue(ROM_BASIS);
-            for (int i = 0; i < mRomDofs; ++i){
-                for (int j = 0; j < mNodalDofs; ++j){
-                    rom_unknowns[i] += nodal_rom_basis(j, i) * rX(node_aux_id * mNodalDofs + j);
-                }
-            }
-        }
-    }
+
+    // Vector ProjectToReducedBasis(
+	// 	const TSystemVectorType& rX,
+	// 	ModelPart::NodesContainerType& rNodes
+	// )
+    // {
+    //     Vector rom_unknowns = ZeroVector(mRomDofs);
+    //     for(const auto& node : rNodes)
+    //     {
+    //         unsigned int node_aux_id = node.GetValue(AUX_ID);
+    //         const auto& nodal_rom_basis = node.GetValue(ROM_BASIS);
+	// 			for (int i = 0; i < mRomDofs; ++i) {
+	// 				for (int j = 0; j < mNodalDofs; ++j) {
+	// 					rom_unknowns[i] += nodal_rom_basis(j, i)*rX(node_aux_id*mNodalDofs + j);
+	// 				}
+	// 			}
+    //     }
+    //     return rom_unknowns;
+	// }
 
     void ProjectToFineBasis(
         const TSystemVectorType &rRomUnkowns,
-        ModelPart::NodesContainerType &rNodes,
-        TSystemVectorType &rX)
+        ModelPart &rModelPart,
+        TSystemVectorType &Dx)
     {
-        TSparseSpace::SetToZero(rX);
-        for (const auto &node : rNodes){
-            unsigned int node_aux_id = node.GetValue(AUX_ID);
-            const auto &nodal_rom_basis = node.GetValue(ROM_BASIS);
-            Vector tmp = prod(nodal_rom_basis, rRomUnkowns);
-            for (unsigned int i = 0; i < tmp.size(); ++i){
-                rX[node_aux_id * mNodalDofs + i] = tmp[i];
-            }
+        const Matrix *current_rom_nodal_basis{nullptr};
+        for (unsigned int k = 0; k<BaseType::mDofSet.size(); k++){
+            auto dof = BaseType::mDofSet.begin() + k;
+            if(k==0)
+                current_rom_nodal_basis = &(rModelPart.pGetNode(dof->Id())->GetValue(ROM_BASIS));
+            else if(dof->Id() != (dof-1)->Id())
+                current_rom_nodal_basis = &(rModelPart.pGetNode(dof->Id())->GetValue(ROM_BASIS));
+            Dx[dof->EquationId()] = inner_prod(  row(  *current_rom_nodal_basis    , MapPhi[dof->GetVariable().Key()]   )     , rRomUnkowns);
         }
     }
 
-    void GetDofValues(const std::vector<DofPointerType> &rDofList, TSystemVectorType &rX)
+    void GetPhiElemental(
+        Matrix &PhiElemental,
+        const Element::DofsVectorType &dofs,
+        const Element::GeometryType &geom)
     {
-        unsigned int i = 0;
-        for (auto &dof : rDofList)
-            rX[i++] = dof->GetSolutionStepValue();
+        auto *current_rom_nodal_basis = &(geom[0].GetValue(ROM_BASIS));

```

Comment:
make this const. you are not changing it

Reply:
done